### PR TITLE
feat: persist popover content on content hover

### PR DESCRIPTION
**Pull request by jason-capsule42**
# Alaska Airlines Pull Request

**Fixes:** #17

## Summary:

Previously the (non-sticky) `auro-popover` content was only displayed when the cursor was over the trigger. This change makes it so that any `auro-popover` which does not have a disconnected trigger (i.e. the trigger element exists inside `<auro-popover >...</auro-popover>` will persist the popover content when the user moves the mouse from the trigger to the content.

## Type of change:

Please delete options that are not relevant.

- [ ] New capability
- [x] Revision of an existing capability
- [ ] Infrastructure change (automation, etc.)
- [ ] Other (please elaborate)

## Checklist:

- [x] My update follows the CONTRIBUTING guidelines of this project
- [x] I have performed a self-review of my own update

**By submitting this Pull Request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

_Pull Requests will be evaluated by their quality of update and whether it is consistent with the goals and values of this project. Any submission is to be considered a conversation between the submitter and the maintainers of this project and may require changes to your submission._

**Thank you for your submission!**<br>
-- Auro Design System Team

### CODE REVIEW

File: src/auro-popover.js [hunk 1]
```diff
@@ -102,8 +102,14 @@ class AuroPopover extends LitElement {
     if (this.sticky) {
       this.trigger.addEventListener('click', handleShow);
     } else {
-      this.trigger.addEventListener('mouseenter', handleShow);
-      this.trigger.addEventListener('mouseleave', handleHide);
+
+      if(this.trigger.parentElement.nodeName === 'AURO-POPOVER') {
```
Comment (geoffrich) [inline src/auro-popover.js:106]: This check seems weird, but I'm not sure if there's a better way.

File: src/auro-popover.js [hunk 2]
```diff
@@ -102,8 +102,14 @@ class AuroPopover extends LitElement {
...
+        this.addEventListener('mouseenter', handleShow);
+        this.addEventListener('mouseleave', handleHide);
+      } else {
+        this.trigger.addEventListener('mouseenter', handleShow);
+        this.trigger.addEventListener('mouseleave', handleHide);
+      }
```
Comment (geoffrich) [inline src/auro-popover.js:112]:
```suggestion
      const element = this.trigger.parentElement.nodeName === 'AURO-POPOVER' ? this : this.trigger;
      element.addEventListener('mouseenter', handleShow);
      element.addEventListener('mouseleave', handleHide);
```

File: src/style.scss [hunk 3]
```diff
@@ -4,89 +4,129 @@
...
   }
 }
 
 .popover {
-  border-radius: var(--auro-border-radius);
-  background: var(--auro-color-background-lightest);
-  box-shadow: -2px 0px 5px 2px rgba(0, 0, 0, 0.08),
-    0px 2px 5px 1px rgba(0, 0, 0, 0.08);
   display: inline-block;
+
```
Comment (geoffrich) [inline src/style.scss:45]: There's a lot of extra newlines and reordering with this PR -- is this aligned with Auro's CSS guidelines?

File: src/style.scss [hunk 4]
```diff
@@ -4,89 +4,129 @@
...
   &:before {
-    transform: rotate(-135deg);
     top: 4px;
     right: -6px;
+
+    transform: rotate(-135deg);
+  }
+}
+
+.popover[data-popper-placement^='top'] {
```
Comment (geoffrich) [inline src/style.scss:86]: This part seems new -- what are these selectors used for?

File: demo/demo.md [hunk 5]
```diff
@@ -65,6 +67,8 @@ In these examples the trigger can be external to the scope of the popover compon
 
 NOTE: The popover element is hidden visually, but not set to `display: none` for accessibility reasons. This is important to know in this setup because the `auro-popover` element may take up physical space depending on your layout solution.
 
+Note that in these examples the popover content will *not* persist as the user moves their mouse off of the trigger and over the popover content.
```
Comment (geoffrich) [inline demo/demo.md:70]: Why is there a difference between this and the slot use case?

**Review by geoffrich — CHANGES_REQUESTED** [review]
A few nits/questions

File: src/auro-popover.js (hunk 1 above)
Comment (jason-capsule42) [reply src/auro-popover.js:106]: Agree, I couldn't think of another way to check that the trigger was not detached.

File: src/style.scss (hunk 4 above)
Comment (jason-capsule42) [reply src/style.scss:86]: The added sections of SCSS are to account for a small gap between the content and the popover. When the user moves the mouse slowly between the two there is a brief moment where the cursor is over neither the trigger or the content so the popover will hide. This block of CSS essentially puts a transparent :before pseudo-element as a bridge between the two to prevent this behavior.

File: src/style.scss (hunk 3 above)
Comment (jason-capsule42) [reply src/style.scss:45]: This is just grouping relevant styles for readability. e.g....

```
    position: absolute;
    bottom: -20px;
    left: 0;

    display: block;

    width: 100%;
    height: 20px;

    content: '';
```

Is much more logically readable than this

```
    display: block;
    position: absolute;
    content: '';
    bottom: -20px;
    left: 0;
    width: 100%;
    height: 20px;

```

Consistent patterning and grouping of the properties makes it a lot easier to read through and parse out what is going on. Also, once the SCSS compiles all the spacing is stripped out anyhow. So it only affects human readability of the code.

File: demo/demo.md (hunk 5 above)
Comment (jason-capsule42) [reply demo/demo.md:70]: By moving the hide/show trigger behavior onto the `auro-popover` element itself rather than the trigger element we essentially get the content hover persistence behavior for free in a manner of speaking.

With a disconnected trigger it's possible to have more of a significant gap in between the trigger and the content depending on how it's implemented. I can't say that anyone would necessarily choose to do that - but nothing stops them from doing it.

Extending the persistence to cover a disconnected trigger would mean something more involved - probably using a timeout to persist the content for a small amount of time after the cursor leaves the trigger and then removing the timeout if the cursor moves over the content before it expires.

This felt like a heavy handed approach, where creating an expectation that persistence meant you had to use a nested trigger seemed more reasonable. What's your opinion on that, @geoffrich?

File: src/auro-popover.js (hunk 2 above)
Comment (jason-capsule42) [reply src/auro-popover.js:112]: I like it - nice, clean adjustment @geoffrich.

File: src/style.scss (hunk 3 above)
Comment (blackfalcon) [reply src/style.scss:45]: Opened a thread for all of us to talk this one out.

File: demo/demo.md (hunk 5 above)
Comment (blackfalcon) [reply demo/demo.md:70]: I am thinking that there is an opportunity here to make this demo more clear. 

1. Popover use cases
2. Using slots, internal to the component
3. Add space around popover
4. Using disconnected elements
5. Do's and don'ts

I don't think we need to add the `works here and not here` lines. When this PR is complete, I want to make updates to all the content on this demo page. 

The thing we can do here is to draw more attention to the idea that if a persisted experience is wanted for disconnected elements, then the `sticky` feature is better suited.

Comment (blackfalcon) [inline demo/demo.md:70]:
```suggestion

```

File: demo/demo.md [hunk 6]
```diff
@@ -20,6 +20,8 @@ Auro popover allows two ways to layout the HTML. Please see the following exampl
 
 In these examples the structure of the HTML uses the slot element to pull the trigger HTML into the scope of the outer shadow DOM element.
 
+Note that in these examples the popover content will persist as the user moves their mouse off of the trigger and over the popover content.
```
Comment (blackfalcon) [inline demo/demo.md:23]:
```suggestion

```

File: package.json [hunk 7]
```diff
@@ -24,8 +24,8 @@
   },
   "peerDependencies": {
     "@alaskaairux/design-tokens": "^3.0.0",
-    "@alaskaairux/webcorestylesheets": "^3.0.0",
     "@alaskaairux/orion-icons": "^2.1.5",
```
Comment (blackfalcon) [inline package.json:27]: Pleae update to use `@alaskaairux/icons`

File: src/style.scss [hunk 8]
```diff
@@ -4,89 +4,129 @@
 // ---------------------------------------------------------------------
 
 // Import Auro tokens
-@import "./node_modules/@alaskaairux/design-tokens/dist/tokens/SCSSVariables";
-@import "./node_modules/@alaskaairux/design-tokens/dist/tokens/SassCustomProperties";
-@import "./node_modules/@alaskaairux/design-tokens/dist/tokens/SCSSVariableMap";
+@import './node_modules/@alaskaairux/design-tokens/dist/tokens/SCSSVariables';
```
Comment (blackfalcon) [inline src/style.scss:7]: Why were these changed from `" "` to `' '`?   Something I go back and forth on as there are preferences on both sides. 

The generator creates a file using double-quotes. If we are to change direction in this, I'd prefer that we set that precedent from the generator, not in an individual PR.

File: src/style.scss (hunk 4 above)
Comment (blackfalcon) [reply src/style.scss:86]:
```suggestion
[data-popper-placement^='top'] {
```

In the scope of a custom element, there can only be one of these. The process of over-qualifying the selector is not needed.

File: src/style.scss [hunk 9]
```diff
@@ -4,89 +4,129 @@
...
+    display: block;
+
+    width: 100%;
+    height: 20px;
+
+    content: '';
+  }
+}
+
+.popover[data-popper-placement^='bottom'] {
```
Comment (blackfalcon) [inline src/style.scss:101]:
```suggestion
data-popper-placement^='bottom'] {
```

**Review by blackfalcon — CHANGES_REQUESTED** [review]

File: src/style.scss (hunk 8 above)
Comment (blackfalcon) [reply src/style.scss:7]: I think I see why you did this, there are some Sass files that are using single quotes. I still stand by the statement above. Let's pick a standard and make sure that everything is that way.

File: src/style.scss (hunk 3 above)
Comment (blackfalcon) [reply src/style.scss:45]: Based on the conversation in Teams, the resolution to this is that this is not a bad pattern, but if this pattern is to be used in a file, that pattern needs to be always used in that file.

File: src/style.scss (hunk 9 above)
Comment (jason-capsule42) [reply src/style.scss:101]: Agreed about the best way to do it. I left it that way because that's how it was done in the previous/original implementation. There are multiple places this selector is so I'll do a patch that cleans it up in all locations.

File: demo/demo.md (hunk 5 above)
Comment (jason-capsule42) [reply demo/demo.md]: Great idea @blackfalcon

File: src/style.scss (hunk 8 above)
Comment (jason-capsule42) [reply src/style.scss:7]: Agreed, let's get into this with the followup on our conversation today to use a formatter consistently.

**Comment by jason-capsule42** [comment]
@blackfalcon This is updated with all the final changes per the review. Just need to resolve the Husky issue. Let me know what you find there. Thanks!

**Review by blackfalcon — DISMISSED** [review]

**Review by blackfalcon — APPROVED** [review]

**Comment by blackfalcon** [comment]
@geoffrich issues have been resolved. PR is approved.
